File: python/agent_beacon/http_url.py
"""Shared validation for outbound HTTP targets."""

from __future__ import annotations

from urllib.parse import urlparse


def require_http_url(value: str, label: str = "URL") -> str:
    """Return a validated absolute HTTP(S) URL without embedded credentials."""
    candidate = value.strip()
    parsed = urlparse(candidate)
    if parsed.scheme not in {"http", "https"} or not parsed.hostname:
        raise ValueError(f"{label} must be an absolute HTTP(S) URL")
    if parsed.username or parsed.password:
        raise ValueError(f"{label} must not contain embedded credentials")
    return candidate


def normalize_http_base_url(value: str, label: str = "Base URL") -> str:
    """Validate a base URL that will have fixed endpoint paths appended."""
    candidate = require_http_url(value, label)
    parsed = urlparse(candidate)
    if parsed.query or parsed.fragment:
        raise ValueError(f"{label} must not contain a query string or fragment")
    return candidate.rstrip("/")

File: src/utils/http-url.ts
/** Validate outbound HTTP targets without adding a runtime dependency. */

import { URL } from 'node:url';

export function requireHttpUrl(value: string, label = 'URL'): URL {
  const candidate = value.trim();
  if (!/^https?:\/\/[^/]/i.test(candidate)) {
    throw new TypeError(`${label} must be an absolute HTTP(S) URL`);
  }

  let parsed: URL;
  try {
    parsed = new URL(candidate);
  } catch (error) {
    throw new TypeError(`${label} must be an absolute HTTP(S) URL`, { cause: error });
  }

  if ((parsed.protocol !== 'http:' && parsed.protocol !== 'https:') || !parsed.hostname) {
    throw new TypeError(`${label} must be an absolute HTTP(S) URL`);
  }
  if (parsed.username || parsed.password) {
    throw new TypeError(`${label} must not contain embedded credentials`);
  }

  return parsed;
}

export function normalizeHttpBaseUrl(value: string, label = 'Base URL'): string {
  const parsed = requireHttpUrl(value, label);
  if (parsed.search || parsed.hash) {
    throw new TypeError(`${label} must not contain a query string or fragment`);
  }
  return parsed.toString().replace(/\/+$/, '');
}
